contracts/randdrop/src/contract.rs

```diff
@@ -0,0 +1,843 @@
+        manager,
+        randdrop_denom,
+        nois_proxy,
+    };
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::default())
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::UpdateConfig {
+            manager,
+            nois_proxy_denom,
+            nois_proxy_amount,
+            nois_proxy_address,
+            randdrop_denom,
+        } => execute_update_config(
+            deps,
+            env,
+            info,
+            manager,
+            randdrop_denom,
+            nois_proxy_denom,
+            nois_proxy_amount,
+            nois_proxy_address,
+        ),
+        ExecuteMsg::RegisterMerkleRoot { merkle_root } => {
+            execute_register_merkle_root(deps, env, info, merkle_root)
+        }
+        // Randdrop should be called by an eligable user to start the process
+        ExecuteMsg::Randdrop { amount, proof } => execute_randdrop(deps, info, amount, proof),
+        // NoisReceive should be called by the proxy contract. The proxy is forwarding the randomness from the nois chain to this contract.
+        ExecuteMsg::NoisReceive { callback } => execute_receive(deps, env, info, callback),
+        ExecuteMsg::WithdrawAll { address } => execute_withdraw_all(deps, env, info, address),
+    }
+}
+
+#[entry_point]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<QueryResponse> {
+    let response = match msg {
+        QueryMsg::IsWinner { address } => to_binary(&query_is_winner(deps, address)?)?,
+        QueryMsg::Config {} => to_binary(&query_config(deps)?)?,
+        QueryMsg::MerkleRoot {} => to_binary(&query_merkle_root(deps)?)?,
+        QueryMsg::IsClaimed { address } => to_binary(&query_is_claimed(deps, address)?)?,
+        QueryMsg::RanddropResults {} => to_binary(&query_results(deps)?)?,
+    };
+    Ok(response)
+}
+
+#[allow(clippy::too_many_arguments)]
+fn execute_update_config(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    manager: Option<String>,
+    randdrop_denom: Option<String>,
+    nois_proxy_denom: Option<String>,
+    nois_proxy_amount: Option<Uint128>,
+    nois_proxy_address: Option<String>,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    // check the calling address is the authorised multisig
+    ensure_eq!(info.sender, config.manager, ContractError::Unauthorized);
+
+    let manager = match manager {
+        Some(ma) => deps.api.addr_validate(&ma)?,
+        None => config.manager,
+    };
+    let randdrop_denom = randdrop_denom.unwrap_or(config.randdrop_denom);
+    let nois_proxy_denom = nois_proxy_denom.unwrap_or(config.nois_proxy.price.denom);
+    let nois_proxy_amount = nois_proxy_amount.unwrap_or(config.nois_proxy.price.amount);
+    let nois_proxy_address = match nois_proxy_address {
+        Some(np) => deps.api.addr_validate(&np)?,
```

Comment:
Here we don't use `InvalidProxyAddress`. So IMO we don't really need the error type.